File: agent-service/src/observationScaling.ts
/**
 * Model-aware observation scaling for native display screenshots.
 *
 * LLM planners emit pixel coordinates in the observation image space we send
 * them. We downscale display captures to that space, then scale coordinates
 * back up before xdotool. Policy is resolved from the configured model so
 * web-vm/desktop stay aligned with the Playwright path and provider limits.
 *
 * The authoritative policy constants live in the shared JSON file
 * ``unify/common/observation_scaling_policy.json``; both this module and the
 * Python ``unify.common.observation_scaling`` read from that single source.
 */

import fs from 'fs';
import path from 'path';

export interface AspectTarget {
  width: number;
  height: number;
}

interface PolicyJson {
  defaultAspectTargets: AspectTarget[];
  providerMaxEdge: Record<string, number>;
  modelObservationMaxEdge: Record<string, number>;
}

function loadPolicy(): PolicyJson {
  const candidates = [
    path.resolve(__dirname, '../../unify/common/observation_scaling_policy.json'),
    path.resolve(__dirname, '../unify/common/observation_scaling_policy.json'),
    path.resolve('/app/unify/common/observation_scaling_policy.json'),
  ];
  for (const candidate of candidates) {
    if (fs.existsSync(candidate)) {
      return JSON.parse(fs.readFileSync(candidate, 'utf-8'));
    }
  }
  throw new Error(
    `observation_scaling_policy.json not found. Searched:\n${candidates.join('\n')}`
  );
}

let _cachedPolicy: PolicyJson | null = null;

function getPolicy(): PolicyJson {
  if (!_cachedPolicy) {
    _cachedPolicy = loadPolicy();
  }
  return _cachedPolicy;
}

/** Same targets as magnitude-core WebHarness (Anthropic computer-use reference). */
export function getDefaultAspectTargets(): AspectTarget[] {
  return getPolicy().defaultAspectTargets;
}

export function getProviderMaxEdge(): Record<string, number> {
  return getPolicy().providerMaxEdge;
}

export function getModelObservationMaxEdge(): Record<string, number> {
  return getPolicy().modelObservationMaxEdge;
}

export interface ObservationScalingPolicy {
  /** Aspect-ratio targets tried before falling back to max-edge scaling. */
  aspectTargets: AspectTarget[];
  /** Hard cap on the longest observation edge for this model/provider. */
  maxEdge: number;
  /** Model id used to resolve this policy (for logs). */
  model: string;
  provider: string;
}

export interface NativeObservationScale {
  displayWidth: number;
  displayHeight: number;
  observationWidth: number;
  observationHeight: number;
  model: string;
  provider: string;
}

/** Parse ``name@provider`` or infer provider from model name. */
export function parseModelProvider(model: string): string {
  const trimmed = model.trim().toLowerCase();
  const at = trimmed.lastIndexOf('@');
  if (at >= 0) {
    return trimmed.slice(at + 1);
  }
  if (trimmed.includes('claude')) return 'anthropic';
  if (trimmed.includes('gpt')) return 'openai';
  if (trimmed.includes('gemini')) return 'google';
  if (trimmed.includes('minimax')) return 'minimax';
  return 'default';
}

function resolveMaxEdge(model: string, provider: string): number {
  const policy = getPolicy();
  const normalized = model.trim().toLowerCase();
  if (policy.modelObservationMaxEdge[normalized] !== undefined) {
    return policy.modelObservationMaxEdge[normalized];
  }
  const envOverride = process.env.UNITY_OBSERVATION_MAX_EDGE?.trim();
  if (envOverride) {
    const parsed = Number(envOverride);
    if (Number.isFinite(parsed) && parsed > 0) {
      return parsed;
    }
  }
  return policy.providerMaxEdge[provider] ?? policy.providerMaxEdge.default;
}

export function resolveObservationScalingPolicy(model: string): ObservationScalingPolicy {
  const provider = parseModelProvider(model);
  return {
    aspectTargets: getDefaultAspectTargets(),
    maxEdge: resolveMaxEdge(model, provider),
    model: model.trim(),
    provider,
  };
}

/**
 * Pick the closest aspect-ratio target for a display size (matches WebHarness).
 * Only scales down; returns null when no target fits.
 */
export function pickAspectTarget(
  displayWidth: number,
  displayHeight: number,
  targets: AspectTarget[],
): AspectTarget | null {
  const ratio = displayWidth / displayHeight;
  for (const target of targets) {
    if (Math.abs(target.width / target.height - ratio) < 0.02) {
      if (target.width < displayWidth) {
        return target;
      }
    }
  }
  return null;
}

function capDimensionsToMaxEdge(
  width: number,
  height: number,
  maxEdge: number,
): { width: number; height: number } {
  const longest = Math.max(width, height);
  if (longest <= maxEdge) {
    return { width, height };
  }
  const factor = maxEdge / longest;
  return {
    width: Math.round(width * factor),
    height: Math.round(height * factor),
  };
}

export function computeNativeObservationScale(
  displayWidth: number,
  displayHeight: number,
  policy: ObservationScalingPolicy,
): NativeObservationScale {
  const aspect = pickAspectTarget(displayWidth, displayHeight, policy.aspectTargets);
  let observationWidth = displayWidth;
  let observationHeight = displayHeight;

  if (aspect) {
    observationWidth = aspect.width;
    observationHeight = aspect.height;
  } else {
    const maxEdge = Math.max(displayWidth, displayHeight);
    if (maxEdge > policy.maxEdge) {
      const factor = policy.maxEdge / maxEdge;
      observationWidth = Math.round(displayWidth * factor);
      observationHeight = Math.round(displayHeight * factor);
    }
  }

  const capped = capDimensionsToMaxEdge(observationWidth, observationHeight, policy.maxEdge);
  observationWidth = capped.width;
  observationHeight = capped.height;

  return {
    displayWidth,
    displayHeight,
    observationWidth,
    observationHeight,
    model: policy.model,
    provider: policy.provider,
  };
}

export function scaleObservationCoordsToDisplay(
  x: number,
  y: number,
  scale: NativeObservationScale,
): { x: number; y: number } {
  if (
    scale.observationWidth === scale.displayWidth
    && scale.observationHeight === scale.displayHeight
  ) {
    return { x, y };
  }
  return {
    x: Math.round(x * (scale.displayWidth / scale.observationWidth)),
    y: Math.round(y * (scale.displayHeight / scale.observationHeight)),
  };
}

File: unify/common/observation_scaling.py
"""Model-aware observation scaling for native display screenshots.

LLM planners emit pixel coordinates in the observation image space we send
them. We downscale display captures to that space, then scale coordinates
back up before xdotool. Policy is resolved from the configured model so
web-vm/desktop stay aligned with the agent-service path and provider limits.

The authoritative policy constants live in the sibling JSON file
``observation_scaling_policy.json``; both this module and the TypeScript
``agent-service/src/observationScaling.ts`` read from that single source.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass
from pathlib import Path
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from PIL import Image as _PILImage

_POLICY_PATH = Path(__file__).with_name("observation_scaling_policy.json")

_policy_cache: dict | None = None


def _load_policy() -> dict:
    global _policy_cache
    if _policy_cache is None:
        with open(_POLICY_PATH) as f:
            _policy_cache = json.load(f)
    return _policy_cache


# ---------------------------------------------------------------------------
# Data structures
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class AspectTarget:
    width: int
    height: int


@dataclass(frozen=True)
class ObservationScalingPolicy:
    """Resolved scaling policy for a given model."""

    aspect_targets: tuple[AspectTarget, ...]
    max_edge: int
    model: str
    provider: str


@dataclass(frozen=True)
class NativeObservationScale:
    """Mapping between display pixels and observation pixels."""

    display_width: int
    display_height: int
    observation_width: int
    observation_height: int
    model: str
    provider: str


# ---------------------------------------------------------------------------
# Policy resolution
# ---------------------------------------------------------------------------


def parse_model_provider(model: str) -> str:
    """Parse ``name@provider`` or infer provider from model name."""
    trimmed = model.strip().lower()
    at = trimmed.rfind("@")
    if at >= 0:
        return trimmed[at + 1 :]
    if "claude" in trimmed:
        return "anthropic"
    if "gpt" in trimmed:
        return "openai"
    if "gemini" in trimmed:
        return "google"
    if "minimax" in trimmed:
        return "minimax"
    return "default"


def _resolve_max_edge(model: str, provider: str) -> int:
    policy = _load_policy()
    normalized = model.strip().lower()

    model_overrides: dict[str, int] = policy["modelObservationMaxEdge"]
    if normalized in model_overrides:
        return model_overrides[normalized]

    env_override = os.environ.get("UNITY_OBSERVATION_MAX_EDGE", "").strip()
    if env_override:
        try:
            parsed = int(env_override)
            if parsed > 0:
                return parsed
        except ValueError:
            pass

    provider_edges: dict[str, int] = policy["providerMaxEdge"]
    return provider_edges.get(provider, provider_edges["default"])


def resolve_observation_scaling_policy(model: str) -> ObservationScalingPolicy:
    """Build the full scaling policy for a model identifier."""
    policy = _load_policy()
    provider = parse_model_provider(model)
    targets = tuple(
        AspectTarget(width=t["width"], height=t["height"])
        for t in policy["defaultAspectTargets"]
    )
    return ObservationScalingPolicy(
        aspect_targets=targets,
        max_edge=_resolve_max_edge(model, provider),
        model=model.strip(),
        provider=provider,
    )


# ---------------------------------------------------------------------------
# Observation model resolution (picks the model driving coordinate space)
# ---------------------------------------------------------------------------


def resolve_observation_model() -> str:
    """Determine the model whose vision limits define the observation space.

    Priority:
      1. CURRENT_ACT_LLM_PROFILE.model (when inside an act() call)
      2. SETTINGS.UNIFY_MODEL (global default)
    """
    from unify.common.act_llm_profiles import CURRENT_ACT_LLM_PROFILE
    from unify.settings import SETTINGS

    try:
        profile = CURRENT_ACT_LLM_PROFILE.get()
        if profile.model:
            return profile.model
    except LookupError:
        pass
    return SETTINGS.UNIFY_MODEL


# ---------------------------------------------------------------------------
# Scaling math (mirrors agent-service/src/observationScaling.ts)
# ---------------------------------------------------------------------------


def pick_aspect_target(
    display_width: int,
    display_height: int,
    targets: tuple[AspectTarget, ...],
) -> AspectTarget | None:
    """Pick the closest aspect-ratio target that scales down from display size."""
    ratio = display_width / display_height
    for target in targets:
        if abs(target.width / target.height - ratio) < 0.02:
            if target.width < display_width:
                return target
    return None


def _cap_dimensions(width: int, height: int, max_edge: int) -> tuple[int, int]:
    longest = max(width, height)
    if longest <= max_edge:
        return width, height
    factor = max_edge / longest
    return round(width * factor), round(height * factor)


def compute_native_observation_scale(
    display_width: int,
    display_height: int,
    policy: ObservationScalingPolicy,
) -> NativeObservationScale:
    """Compute the observation image dimensions for a given display size."""
    aspect = pick_aspect_target(display_width, display_height, policy.aspect_targets)

    if aspect:
        obs_w, obs_h = aspect.width, aspect.height
    else:
        longest = max(display_width, display_height)
        if longest > policy.max_edge:
            factor = policy.max_edge / longest
            obs_w = round(display_width * factor)
            obs_h = round(display_height * factor)
        else:
            obs_w, obs_h = display_width, display_height

    obs_w, obs_h = _cap_dimensions(obs_w, obs_h, policy.max_edge)

    return NativeObservationScale(
        display_width=display_width,
        display_height=display_height,
        observation_width=obs_w,
        observation_height=obs_h,
        model=policy.model,
        provider=policy.provider,
    )


def scale_observation_coords_to_display(
    x: int,
    y: int,
    scale: NativeObservationScale,
) -> tuple[int, int]:
    """Map LLM-emitted observation coordinates back to display pixels."""
    if (
        scale.observation_width == scale.display_width
        and scale.observation_height == scale.display_height
    ):
        return x, y
    return (
        round(x * (scale.display_width / scale.observation_width)),
        round(y * (scale.display_height / scale.observation_height)),
    )


# ---------------------------------------------------------------------------
# Image fitting (used by capture.py display path)
# ---------------------------------------------------------------------------


def fit_image_to_observation_space(
    img: "_PILImage.Image",
    model: str | None = None,
) -> "_PILImage.Image":
    """Resize a PIL Image to the observation space for the given model.

    Idempotent: returns the image unchanged if it already fits within the
    observation dimensions.
    """
    from PIL import Image as _Image

    if model is None:
        model = resolve_observation_model()

    policy = resolve_observation_scaling_policy(model)
    w, h = img.size
    scale = compute_native_observation_scale(w, h, policy)

    if scale.observation_width == w and scale.observation_height == h:
        return img

    return img.resize(
        (scale.observation_width, scale.observation_height),
        _Image.LANCZOS,
    )
